File: internal/hub/lexorank/lexorank_test.go
package lexorank

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestFirst(t *testing.T) {
	f := First()
	assert.NotEmpty(t, f)
}

func TestAfter(t *testing.T) {
	a := First()
	b := After(a)
	assert.Greater(t, b, a)

	c := After(b)
	assert.Greater(t, c, b)
}

func TestMid(t *testing.T) {
	tests := []struct {
		name string
		a, b string
	}{
		{"both empty", "", ""},
		{"a empty", "", "n"},
		{"b empty", "n", ""},
		{"simple", "a", "z"},
		{"adjacent", "a", "b"},
		{"close", "a", "c"},
		{"mid range", "m", "o"},
		{"multi char", "abc", "abd"},
		{"different lengths", "a", "bb"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := Mid(tt.a, tt.b)
			assert.NotEmpty(t, result)
			if tt.a != "" {
				assert.Greater(t, result, tt.a)
			}
			if tt.b != "" {
				assert.Less(t, result, tt.b)
			}
		})
	}
}

func TestMidSequential(t *testing.T) {
	// Generate a sequence of ranks by repeatedly inserting at the end.
	ranks := []string{First()}
	for i := 0; i < 20; i++ {
		next := After(ranks[len(ranks)-1])
		assert.Greater(t, next, ranks[len(ranks)-1], "iteration %d", i)
		ranks = append(ranks, next)
	}

	// Verify all ranks are strictly increasing.
	for i := 1; i < len(ranks); i++ {
		assert.Greater(t, ranks[i], ranks[i-1], "ranks[%d]=%q should be > ranks[%d]=%q", i, ranks[i], i-1, ranks[i-1])
	}
}

func TestMidInsertBetween(t *testing.T) {
	// Create two ranks and repeatedly insert between them.
	a := "b"
	b := "y"
	for i := 0; i < 20; i++ {
		m := Mid(a, b)
		assert.Greater(t, m, a, "iteration %d", i)
		assert.Less(t, m, b, "iteration %d", i)
		// Alternate inserting before and after the midpoint.
		if i%2 == 0 {
			b = m
		} else {
			a = m
		}
	}
}

func TestBefore(t *testing.T) {
	s := "n"
	b := before(s)
	assert.Less(t, b, s)

	// Repeatedly insert before.
	for i := 0; i < 10; i++ {
		prev := before(b)
		assert.Less(t, prev, b, "iteration %d", i)
		b = prev
	}
}

File: frontend/tests/unit/lib/lexorank.test.ts
import { describe, expect, it } from 'vitest'
import { after, first, mid } from '~/lib/lexorank'

describe('lexorank', () => {
  describe('first', () => {
    it('should return a non-empty string', () => {
      expect(first()).toBeTruthy()
      expect(first()).toBe('n')
    })
  })

  describe('after', () => {
    it('should return a rank after the given string', () => {
      const a = first()
      const b = after(a)
      expect(b > a).toBe(true)
    })

    it('should produce strictly increasing sequence', () => {
      let prev = first()
      for (let i = 0; i < 10; i++) {
        const next = after(prev)
        expect(next > prev).toBe(true)
        prev = next
      }
    })
  })

  describe('mid', () => {
    it('should return first() when both args are empty', () => {
      expect(mid('', '')).toBe(first())
    })

    it('should return rank before b when a is empty', () => {
      const b = 'n'
      const result = mid('', b)
      expect(result < b).toBe(true)
      expect(result.length).toBeGreaterThan(0)
    })

    it('should return rank after a when b is empty', () => {
      const a = 'n'
      const result = mid(a, '')
      expect(result > a).toBe(true)
    })

    it('should return rank between a and b', () => {
      const a = 'b'
      const b = 'z'
      const result = mid(a, b)
      expect(result > a).toBe(true)
      expect(result < b).toBe(true)
    })

    it('should handle adjacent characters', () => {
      const a = 'a'
      const b = 'b'
      const result = mid(a, b)
      expect(result > a).toBe(true)
      expect(result < b).toBe(true)
    })

    it('should produce valid mid for sequential insertions', () => {
      const ranks: string[] = [first()]
      for (let i = 0; i < 20; i++) {
        const next = mid(ranks[ranks.length - 1], '')
        expect(next > ranks[ranks.length - 1]).toBe(true)
        ranks.push(next)
      }

      // All ranks should be strictly ordered.
      for (let i = 1; i < ranks.length; i++) {
        expect(ranks[i] > ranks[i - 1]).toBe(true)
      }
    })

    it('should produce valid mid for insertions between bounds', () => {
      const lo = 'a'
      let hi = 'z'
      for (let i = 0; i < 20; i++) {
        const m = mid(lo, hi)
        expect(m > lo).toBe(true)
        expect(m < hi).toBe(true)
        hi = m
      }
    })

    it('should handle equal strings', () => {
      const result = mid('nn', 'nn')
      expect(result > 'nn').toBe(true)
    })
  })
})
